UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs

```diff
@@ -199,6 +208,192 @@ public static MatrixInfo AtPoint(Vector3Int worldPos, bool isServer)
+		LayerTypeSelection layerMask, LayerMask? Layermask2D = null, Vector3? WorldTo = null)
+	{
+
+
+		Worldorigin.z = 0;
+
+
+		//TODO RRT
+		//get to  from vector
+		CustomPhysicsHit? ClosestHit = null;
+		CustomPhysicsHit? Checkhit = null;
+		// Vector3 Localdorigin = Vector3.zero;
+		// Vector3 LocalTo = Vector3.zero;
+
+		if (WorldTo == null)
+		{
+			WorldTo = Worldorigin + (Vector3) (direction.normalized * distance);
+		}
+
+		if (direction.x == 0 && direction.y == 0)
+		{
+			direction = (WorldTo.Value - Worldorigin).normalized;
+		}
+
+		if (layerMask != LayerTypeSelection.None)
+		{
+			for (var i = Instance.ActiveMatrices.Count - 1; i >= 0; i--)
+			{
+				MatrixInfo mat = Instance.ActiveMatrices[i];
+				//if (mat.Matrix == Instance.spaceMatrix) continue;
+				if (LineIntersectsRect(Worldorigin, (Vector2) WorldTo, mat.WorldBounds))
+				{
+					Checkhit = mat.MetaTileMap.Raycast(Worldorigin.ToLocal(mat.Matrix), Vector2.zero, distance,
+						layerMask,
+						WorldTo.Value.ToLocal(mat.Matrix));
+
+
+					if (Checkhit != null)
+					{
+						if (ClosestHit != null)
+						{
+							if (ClosestHit.Value.Distance < Checkhit.Value.Distance)
+							{
+								ClosestHit = Checkhit;
+							}
+						}
+						else
+						{
+							ClosestHit = Checkhit;
+						}
+					}
+				}
+			}
+		}
+
+		if (Layermask2D != null)
+		{
+			var Hit2D = Physics2D.Raycast(Worldorigin, direction.normalized, distance, Layermask2D.Value);
+			if (ClosestHit != null)
+			{
+				if (Hit2D.distance != 0 && ClosestHit.Value.Distance > Hit2D.distance)
+				{
+					ClosestHit = new CustomPhysicsHit(Hit2D);
+				}
+			}
+			else
+			{
+				ClosestHit = new CustomPhysicsHit(Hit2D);
+			}
+
+		}
+
+		if (ClosestHit == null)
+		{
+			ClosestHit = new CustomPhysicsHit();
+		}
+
+		return ClosestHit.Value;
+	}
+
+	public struct CustomPhysicsHit
+	{
+		public Vector3 TileHitWorld;
+		public Vector3 HitWorld;
+		public Vector2 Normal;
+		public float Distance;
+		public CollisionHit CollisionHit;
+		public bool ItHit;
+
+		public CustomPhysicsHit(bool unneededbool = false)
+		{
+			TileHitWorld = Vector3.zero;
+			HitWorld = Vector3.zero;
+			Normal = Vector2.zero;
+			Distance = 9999999999f;
+			CollisionHit = new CollisionHit(null, null);
+			ItHit = false;
+		}
+
+		public CustomPhysicsHit(Vector3 _TileHitWorld, Vector3 _HitWorld, Vector2 _Normal, float _RelativeHit,
+			TileLocation TileHit)
+		{
+			TileHitWorld = _TileHitWorld;
+			HitWorld = _HitWorld;
+			Normal = _Normal;
+			Distance = _RelativeHit;
+			CollisionHit = new CollisionHit(null, TileHit);
+			ItHit = TileHit != null;
+		}
+
+		public CustomPhysicsHit(RaycastHit2D raycastHit2D)
+		{
+			TileHitWorld = raycastHit2D.point;
+			HitWorld = raycastHit2D.point;
+			Normal = raycastHit2D.normal;
+			Distance = raycastHit2D.distance;
+			CollisionHit = new CollisionHit(raycastHit2D.collider?.gameObject, null);
+			ItHit = raycastHit2D.collider?.gameObject != null;
+		}
+	}
+
+	public struct CollisionHit
```

Comment:
Codacy found an issue: [Implement 'IEquatable<T>' in value type 'CollisionHit'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6319358)